src/Arcus.Messaging.Abstractions.ServiceBus/AzureServiceBusMessageContext.cs

```diff
@@ -135,8 +157,156 @@ public async Task DeadLetterMessageAsync(string deadLetterReason, string deadLet
+            }
+
+            public Task DeadLetterMessageAsync(string deadLetterReason, string deadLetterErrorDescription, IDictionary<string, object> newMessageProperties, CancellationToken cancellationToken)
+            {
+                return _receiver.DeadLetterMessageAsync(Message, newMessageProperties, deadLetterReason, deadLetterErrorDescription, cancellationToken);
+            }
+
+            public Task AbandonMessageAsync(IDictionary<string, object> newMessageProperties, CancellationToken cancellationToken)
+            {
+                return _receiver.AbandonMessageAsync(Message, newMessageProperties, cancellationToken);
+            }
         }
+
+        private sealed class MessageSettlementViaSessionEventArgs : IMessageSettlementStrategy
+        {
+            private readonly ProcessSessionMessageEventArgs _args;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="MessageSettlementViaSessionEventArgs"/> class.
+            /// </summary>
+            /// <param name="processSessionMessageEventArgs">The <see cref="ProcessSessionMessageEventArgs"/> that contains the information
+            /// about the ServiceBus message that has been received.</param>
+            public MessageSettlementViaSessionEventArgs(ProcessSessionMessageEventArgs processSessionMessageEventArgs)
+            {
+                _args = processSessionMessageEventArgs;
+                Message = _args.Message;
+                FullyQualifiedNamespace = _args.FullyQualifiedNamespace;
+                EntityPath = _args.EntityPath;
+            }
+
+            public string FullyQualifiedNamespace { get; }
+            public string EntityPath { get; }
+            public ServiceBusReceivedMessage Message { get; }
+
+            public Task CompleteMessageAsync(CancellationToken cancellationToken)
+            {
+                return _args.CompleteMessageAsync(_args.Message, cancellationToken);
+            }
+
+            public Task DeadLetterMessageAsync(string deadLetterReason, string deadLetterErrorDescription,
+                CancellationToken cancellationToken)
+            {
+                return DeadLetterMessageAsync(deadLetterReason, deadLetterErrorDescription, newMessageProperties: null, cancellationToken);
+            }
+
+            public Task DeadLetterMessageAsync(string deadLetterReason, string deadLetterErrorDescription, IDictionary<string, object> newMessageProperties, CancellationToken cancellationToken)
+            {
+                if (newMessageProperties != null)
+                {
+                    var propertiesToModify = newMessageProperties.ToDictionary();
+                    return _args.DeadLetterMessageAsync(_args.Message, propertiesToModify, deadLetterReason, deadLetterErrorDescription, cancellationToken);
+                }
+
+                return _args.DeadLetterMessageAsync(_args.Message, deadLetterReason, deadLetterErrorDescription, cancellationToken);
```

Comment:
This is a strange construction, but this is due to some weird implementation choice in the ServiceBus SDK I believe.
The `ProcessSessionMessageEventArgs.DeadLetterMessageAsync` has [these](https://learn.microsoft.com/en-us/dotnet/api/azure.messaging.servicebus.processsessionmessageeventargs.deadlettermessageasync?view=azure-dotnet) overloads.
Note the fact that one method accepts a `Dictionary` and another one accepts an `IDictionary` (that allows nulls, as the default value for that parameter is null).
When the `DeadLetterMessageAsync` method that accepts a `Dictionary` is called, and you pass `null` to that parameter, a `NullReferenceException` is being thrown.